File: test-suite/src/test/java/io/micronaut/security/docs/securityrule/secured/SecuredTest.java
package io.micronaut.security.docs.securityrule.secured;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.security.authentication.AuthenticationFailureReason;
import io.micronaut.security.authentication.AuthenticationRequest;
import io.micronaut.security.authentication.AuthenticationResponse;
import io.micronaut.security.authentication.provider.HttpRequestAuthenticationProvider;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@Property(name = "spec.name", value = "SecuredTest")
@MicronautTest
class SecuredTest {

    @Test
    void anonymousEndpointCanBeAccessedWithoutAuthentication(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        assertDoesNotThrow(() -> client.exchange(HttpRequest.GET("/example/anonymous")));
    }

    @Test
    void authenticatedEndpointRequiresAnAuthenticatedUser(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        Executable e = () -> client.exchange(HttpRequest.GET("/example/authenticated"));
        HttpClientResponseException ex = assertThrows(HttpClientResponseException.class, e);
        assertEquals(HttpStatus.UNAUTHORIZED, ex.getStatus());
        assertDoesNotThrow(() -> client.exchange(HttpRequest.GET("/example/authenticated").basicAuth("user", "password")));
    }

    @Test
    void adminEndpointRequiresOneOfTheRoles(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        Executable e = () -> client.exchange(HttpRequest.GET("/example/admin"));
        HttpClientResponseException ex = assertThrows(HttpClientResponseException.class, e);
        assertEquals(HttpStatus.UNAUTHORIZED, ex.getStatus());
        e = () -> client.exchange(HttpRequest.GET("/example/admin").basicAuth("user", "password"));
        ex = assertThrows(HttpClientResponseException.class, e);
        assertEquals(HttpStatus.FORBIDDEN, ex.getStatus());
        assertDoesNotThrow(() -> client.exchange(HttpRequest.GET("/example/admin").basicAuth("admin", "password")));
    }

    @Requires(property = "spec.name", value = "SecuredTest")
    @Singleton
    static class AuthenticationProviderUserPassword<B> implements HttpRequestAuthenticationProvider<B> {
        @Override
        public AuthenticationResponse authenticate(HttpRequest<B> requestContext, AuthenticationRequest<String, String> authRequest) {
            if (authRequest.getIdentity().equals("user")) {
                return AuthenticationResponse.success("user");
            }
            if (authRequest.getIdentity().equals("admin")) {
                return AuthenticationResponse.success("admin", List.of("ROLE_ADMIN"));
            }
            return AuthenticationResponse.failure(AuthenticationFailureReason.USER_NOT_FOUND);
        }
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/docs/securityrule/secured/SecuredTest.kt
package io.micronaut.security.docs.securityrule.secured

import io.micronaut.context.annotation.Property
import io.micronaut.context.annotation.Requires
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.http.client.exceptions.HttpClientResponseException
import io.micronaut.security.authentication.AuthenticationFailureReason
import io.micronaut.security.authentication.AuthenticationRequest
import io.micronaut.security.authentication.AuthenticationResponse
import io.micronaut.security.authentication.provider.HttpRequestAuthenticationProvider
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Singleton
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

@Property(name = "spec.name", value = "SecuredTest")
@MicronautTest
internal class SecuredTest {

    @Test
    fun anonymousEndpointCanBeAccessedWithoutAuthentication(@Client("/") httpClient: HttpClient) {
        val client = httpClient.toBlocking()
        assertDoesNotThrow { client.exchange<Any, Any>(HttpRequest.GET("/example/anonymous")) }
    }

    @Test
    fun authenticatedEndpointRequiresAnAuthenticatedUser(@Client("/") httpClient: HttpClient) {
        val client = httpClient.toBlocking()
        val ex = assertThrows(HttpClientResponseException::class.java) {
            client.exchange<Any, Any>(HttpRequest.GET("/example/authenticated"))
        }
        assertEquals(HttpStatus.UNAUTHORIZED, ex.status)
        assertDoesNotThrow { client.exchange<Any, Any>(HttpRequest.GET<Any>("/example/authenticated").basicAuth("user", "password")) }
    }

    @Test
    fun adminEndpointRequiresOneOfTheRoles(@Client("/") httpClient: HttpClient) {
        val client = httpClient.toBlocking()
        var ex = assertThrows(HttpClientResponseException::class.java) {
            client.exchange<Any, Any>(HttpRequest.GET("/example/admin"))
        }
        assertEquals(HttpStatus.UNAUTHORIZED, ex.status)
        ex = assertThrows(HttpClientResponseException::class.java) {
            client.exchange<Any, Any>(HttpRequest.GET<Any>("/example/admin").basicAuth("user", "password"))
        }
        assertEquals(HttpStatus.FORBIDDEN, ex.status)
        assertDoesNotThrow { client.exchange<Any, Any>(HttpRequest.GET<Any>("/example/admin").basicAuth("admin", "password")) }
    }

    @Requires(property = "spec.name", value = "SecuredTest")
    @Singleton
    internal class AuthenticationProviderUserPassword<B : Any> : HttpRequestAuthenticationProvider<B> {
        override fun authenticate(requestContext: HttpRequest<B>?, authRequest: AuthenticationRequest<String, String>): AuthenticationResponse =
            when (authRequest.identity) {
                "user" -> AuthenticationResponse.success("user")
                "admin" -> AuthenticationResponse.success("admin", listOf("ROLE_ADMIN"))
                else -> AuthenticationResponse.failure(AuthenticationFailureReason.USER_NOT_FOUND)
            }
    }
}
